include/flang/lower/OpBuilder.h

```diff
@@ -115,24 +116,74 @@ class CharacterOpsBuilder : public OpBuilderWrapper {
   CharacterOpsBuilder(OpBuilderWrapper &b) : OpBuilderWrapper{b} {}
   /// Interchange format to avoid inserting unbox/embox everywhere while
   /// evaluating character expressions.
-  struct CharValue {
-    fir::ReferenceType getReferenceType();
-    fir::CharacterType getCharacterType();
 
-    mlir::Value reference;
+  /// A readable character entity
+  struct CharValue {
+    /// Get fir.char<kind> type with the same kind as value.
+    fir::CharacterType getCharacterType() const;
+    /// Get fir.ref<fir.char<kind>> type.
+    fir::ReferenceType getReferenceType() const;
+    /// Get fir.sequence<? x fir.char<kind>> type.
+    fir::SequenceType getSequenceType() const;
+    /// Get fir.ref<fir.sequence<? x fir.char<kind>>> type.
+    fir::ReferenceType getSequenceRefType() const;
+
+    /// value is a fir.ref<...>
+    bool isRef() const;
+    /// value is not a fir.ref<...>
+    bool isValue() const { return !isRef(); };
+
+    /// Value must be of type:
+    /// - fir.char<kind>
+    /// - fir.sequence<? x fir.char<kind>> type.
+    /// - fir.ref<fir.char<kind>>
+    /// - fir.ref<fir.sequence<? x fir.char<kind>>> type.
+    mlir::Value value;
     mlir::Value len;
   };
 
+  /// A readable and writable character entity.
+  struct CharRef : CharValue {
+    /// value must be of type:
+    /// - fir.ref<fir.char<kind>>
+    /// - fir.ref<fir.sequence<? x fir.char<kind>>> type.
+  };
+
+  /// A readable character entity that can be indexed.
+  struct AddressableValue : CharValue {
+    /// value must be of type:
+    /// - fir.sequence<? x fir.char<kind>> type.
+    /// - fir.ref<fir.sequence<? x fir.char<kind>>> type.
+  };
+
+  /// A readable and writable character entity that can be indexed.
+  struct AddressableRef : AddressableValue {
+    /// value can be of type:
+    /// - fir.ref<fir.sequence<? x fir.char<kind>>> type.
+  };
```

Comment:
Not sure about the extra level of indirection here ...

Reply:
Yeah.. it was to provide a way to enforce at f18 compile time that we only attempt to write on character reference. `mlir::Value` type erasure is nice, but sometimes I think it may also hide quit a lot of bugs that will not appear until a weird test hits an assert.  That may be a bit overkill tough, I'll throw that and rely on asserts/mlir errors for now.

Reply:
It seems not a perfect argument to say problems are "hidden" with assertions. FIR is strongly typed and the IR is verified. With an assertion a bug is exposed at run-time where it happens.

Reply:
What I meant is that you can write:
```
auto char1{builder.create<fir::StringLitOp>(...)};
auto char2{builder.create<fir::StringLitOp>(...)};
auto add{builder.create<fir::AddOp>(..., char1, char2)}; // That is obviously always wrong.
```
f18 will be compiled just fine, and until we run f18 with on a program that test this piece of code, the assert checking types will not blow up. That is fine, but it really means we need a lot of tests to validate our code. Providing interfaces that allow some *static* type checking sometimes help in my opinion.

Reply:
Yes, I knew you meant compile-time failure.